src/main/utility/interval_map.rs

```diff
@@ -0,0 +1,610 @@
+    pub fn iter(&self) -> ItemIter<V> {
+        ItemIter { map: self, i: 0 }
+    }
+
+    /// Mutates the map so that the given range maps to nothing, modifying and removing intervals
+    /// as needed. Returns what mutations were performed, including the values of any
+    /// completely-removed intervals. If an interval is split (e.g. by inserting [5,6] into
+    /// [0,10]), its value will be cloned.
+    ///
+    /// Returned mutations are ordered by their original interval values.
+    pub fn clear(&mut self, begin: usize, end: usize) -> Vec<Mutation<V>> {
+        self.splice(begin, end, None)
+    }
+
+    /// Insert range from `begin` to `end`, inclusive, mapping that range to `val`.  Existing
+    /// contents of that range are cleared, and mutations returned, as for `clear`.
+    pub fn insert(&mut self, begin: usize, end: usize, val: V) -> Vec<Mutation<V>> {
+        self.splice(begin, end, Some(val))
+    }
+
+    // Splices zero or one value into the given interval.
+    fn splice(&mut self, begin: usize, end: usize, val: Option<V>) -> Vec<Mutation<V>> {
+        assert!(begin <= end);
+        let mut mutations = Vec::new();
+        let mut begins_insertions = Vec::new();
+        let mut ends_insertions = Vec::new();
+        let mut vals_insertions = Vec::new();
+        if let Some(v) = val {
+            begins_insertions.push(begin);
+            ends_insertions.push(end);
+            vals_insertions.push(v);
+        };
+
+        // We're eventually going to call Vec::splice on our vectors, and
+        // this will be the starting index.
+        let splice_start = match self.begins.binary_search(&begin) {
+            Ok(i) | Err(i) => i,
+        };
+
+        // The eventual splice will be with a non-inclusive end-point. i.e. we start with
+        // replacing no items, but will expand this if there are intervals we need to remove.
+        let mut splice_end = splice_start;
+
+        // Check whether there's an interval before the splice point,
+        // and if so whether it overlaps.
+        if splice_start > 0 && self.ends[splice_start - 1] >= begin {
+            let i = splice_start - 1;
+
+            // If it ends after the end of our interval, we need to split it.
+            if self.ends[i] > end {
+                let old = Interval::new(self.begins[i], self.ends[i]);
+                let new1 = Interval::new(self.begins[i], begin - 1);
+                let new2 = Interval::new(end + 1, self.ends[i]);
+
+                // Truncate the existing interval.
+                self.ends[i] = new1.end();
+
+                // Create a new interval, starting after the insertion interval.
+                begins_insertions.push(new2.begin());
+                ends_insertions.push(new2.end());
+                vals_insertions.push(self.vals[i].clone());
+                mutations.push(Mutation::Split(old, new1, new2));
+            } else {
```

Comment:
A nitpick, but I think it would help readability if the shorter case came first, sort of as an introduction to all of the variables.

Reply:
Done